Provide better explanations of dot context

rehevkor5 opened the pull request:
- See https://github.com/helm/helm/issues/11232 and https://github.com/helm/helm/issues/7915 for examples of confusion the current explanation/wording has caused.
- Replaced inconsistent and unexplained terminology: "root context", "root scope", "top-level scope", "parent scope".
- Go templates/gomplate, on which Helm relies, uses the words "dot" and "context", not "scope". The word "scope" is better used to explain when variables go out of scope, not to explain the value of the context variables `.` and `$`.
- Fixed inaccurate or misleading statements such as: $ "does not change", and $ "will always point to the root context".
- This sentence: "Recall that `.` is a reference to _the current scope_." is not useful - this was not explained elsewhere
- Removed vague sentences that don't add meaning, such as "Values are namespaced, but namespaces are pruned."

Signed-off-by: Shannon Carey <rehevkor5@gmail.com>

rehevkor5 committed:
Provide better explanations of dot context

- See https://github.com/helm/helm/issues/11232 and
https://github.com/helm/helm/issues/7915 for examples of confusion the
current explanation/wording has caused.
- Replaced inconsistent and unexplained terminology: "root context",
"root scope", "top-level scope", "parent scope".
- Go templates/gomplate, on which Helm relies, uses the words "dot" and
"context", not "scope". The word "scope" is better used to explain when
variables go out of scope, not to explain the value of the context
variables `.` and `$`.
- Fixed inaccurate or misleading statements such as:
 $ "does not change", and $ "will always point to the root context".
- This sentence: "Recall that `.` is a reference to _the current
scope_." is not useful - this was not explained elsewhere
- Removed vague sentences that don't add meaning, such as "Values are
namespaced, but namespaces are pruned."

Signed-off-by: Shannon Carey <rehevkor5@gmail.com>

miles-w-3 commented:
This is a great set of changes, I didn't know that `with` could act as a conditional operator

rehevkor5 commented:
Anything I can do to move this forward?

VannTen commented on content/en/docs/chart_template_guide/named_templates.md:
```diff
@@ -190,17 +190,18 @@ metadata:
-No scope was passed in, so within the template we cannot access anything in `.`.
-This is easy enough to fix, though. We simply pass a scope to the template:
+No pipeline value was provided to its second parameter, so within the template `.` is empty.
+`$.` will therefore also be empty.
```
I think it should be `$` without the dot here.

rehevkor5 replied:
I see what you mean. I guess I got caught by the inconsistency that `.` is analogous to `$` but `.Foo` is analogous to `$.Foo` (extra dot appears from thin air) instead of `$Foo`. Something about `.` is magical I guess.

```
% echo "{{ . }}" | gomplate
map[]
% echo "{{ $. }}" | gomplate
16:41:50 ERR  error="template: -:1: unexpected <.> in operand"
```

Fixing...

rehevkor5 committed:
Dot "." is equivalent to "$" not "$."

- For example:
% echo "{{ . }}" | gomplate
map[]
% echo "{{ $. }}" | gomplate
16:41:50 ERR  error="template: -:1: unexpected <.> in operand"

Signed-off-by: Shannon Carey <rehevkor5@gmail.com>

yurrriq commented:
I wonder if it would be better to eschew the redundant parts (e.g. `if`, `with`, `range`) in favor of a link to https://pkg.go.dev/text/template, which explains the actions nicely.

rehevkor5 commented:
> I wonder if it would be better to eschew the redundant parts (e.g. if, with, range) in favor of a link to https://pkg.go.dev/text/template, which explains the actions nicely.

You might be right, though I didn't want to make this change any more difficult to accept than it probably already is.

If such a alteration would make this PR more likely to be accepted, I would be happy to do it.

rehevkor5 commented:
Maybe someone like @yxxhero can review?

rehevkor5 commented:
gg
